library/Icingadb/Common/Macros.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+/* Icinga DB Web | (c) 2021 Icinga GmbH | GPLv2 */
+
+namespace Icinga\Module\Icingadb\Common;
+
+use Icinga\Application\Logger;
+use Icinga\Module\Icingadb\Model\Host;
+use Icinga\Module\Icingadb\Model\Service;
+
+trait Macros
+{
+    /**
+     * Get the given string with macros being resolved
+     *
+     * @param string       $input  The string in which to look for macros
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function expandMacros(string $input, $object): string
+    {
+        if (preg_match_all('@\$([^\$\s]+)\$@', $input, $matches)) {
+            foreach ($matches[1] as $key => $value) {
+                $newValue = $this->resolveMacro($value, $object);
+                if ($newValue !== $value) {
+                    $input = str_replace($matches[0][$key], $newValue, $input);
+                }
+            }
+        }
+
+        return $input;
+    }
+
+    /**
+     * Resolve a macro based on the given object
+     *
+     * @param string       $macro  The macro to resolve
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function resolveMacro(string $macro, $object): string
+    {
+        if ($object instanceof Host) {
+            $objectType = 'host';
+        } else {
+            $objectType = 'service';
+        }
+
+        $macroType = null;
+        $isCustomVar = false;
+        if (preg_match('/^((host|service)[.])?vars[.](.+)/i', $macro, $matches)) {
+            if (! empty($matches[2])) {
+                $macroType = $matches[2];
+            } else {
+                $macroType = $objectType;
+            }
+
+            $path = $matches[3];
+            $isCustomVar = true;
+        } elseif (preg_match('/^(host|service)[.](.+)/i', $macro, $matches)) {
+            $macroType = $matches[1];
+            $properties = explode('.', $matches[2]);
+        }
+
+        try {
+            if ($macroType && ($macroType === $objectType || $objectType === 'service')) {
+                if ($macroType !== $objectType) {
+                    $value = $object->$macroType;
+                } else {
+                    $value = $object;
+                }
+
+                if ($isCustomVar) {
+                    $value = $value->vars[$path];
+                } else {
+                    do {
+                        $column = array_shift($properties);
+                        $value = $value->$column;
```

Comment:
This allows to access `host.hostgroup` for example. That's bad since it cannot be resolved to a useful value and might issue a query. Don't resolve such macros that result in an access to properties that have `Query`, `ResultSet` or `array` instance as value.